pkg/testing/ess/statful_provisioner.go

```diff
@@ -142,12 +145,72 @@ func (p *statefulProvisioner) Delete(ctx context.Context, stack common.Stack) er
 	return p.client.ShutdownDeployment(ctx, deploymentID)
 }
 
-func (p *statefulProvisioner) createDeployment(ctx context.Context, r common.StackRequest, tags map[string]string) (*CreateDeploymentResponse, error) {
+// Upgrade upgrades a stack to a new version.
+func (p *StatefulProvisioner) Upgrade(ctx context.Context, stack common.Stack, newVersion *version.ParsedSemVer) error {
+	deploymentID, err := p.getDeploymentID(stack)
+	if err != nil {
+		return fmt.Errorf("failed to get deployment ID from the stack: %w", err)
+	}
+
+	p.logger.Logf("Upgrading cloud stack %s [stack_id: %s, deployment_id: %s] to version %s", stack.Version, stack.ID, deploymentID, newVersion)
+
+	// allow up to 10 minutes for request
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
+	defer cancel()
+
+	err = p.client.UpgradeDeployment(ctx, deploymentID, newVersion.String())
+	if err != nil {
+		return fmt.Errorf("failed to upgrade cloud stack %s [stack_id: %s, deployment_id: %s] to version %s: %w", stack.Version, stack.ID, deploymentID, newVersion.String(), err)
+	}
+
+	return nil
+}
+
+// AvailableVersions returns the stack versions available in the ECH region.
+func (p *StatefulProvisioner) AvailableVersions() ([]*version.ParsedSemVer, error) {
+	versionsApiUrl, err := url.JoinPath("regions", p.cfg.Region, "stack/versions?show_deleted=false&show_unusable=false")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create ECH versions API URL: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	resp, err := p.client.doGet(ctx, versionsApiUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get versions from ECH for region [%s]: %w", p.cfg.Region, err)
+	}
+	defer resp.Body.Close()
+
+	var stacks struct {
+		Stacks []struct {
+			Version string `json:"version"`
+		} `json:"stacks"`
+	}
+	decoder := json.NewDecoder(resp.Body)
+	err = decoder.Decode(&stacks)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse ECH versions API response from region [%s]: %w", p.cfg.Region, err)
+	}
+
+	var versions []*version.ParsedSemVer
+	for _, stack := range stacks.Stacks {
+		ver, err := version.ParseVersion(stack.Version)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse stack version [%s]: %w", stack.Version, err)
+		}
+		versions = append(versions, ver)
+	}
+
+	return versions, nil
+}
+
+func (p *StatefulProvisioner) createDeployment(ctx context.Context, r common.StackRequest, tags map[string]string) (*CreateDeploymentResponse, error) {
 	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
 	defer cancel()
 
 	p.logger.Logf("Creating cloud stack %s [stack_id: %s]", r.Version, r.ID)
-	name := fmt.Sprintf("%s-%s", strings.Replace(p.cfg.Identifier, ".", "-", -1), r.ID)
+	name := fmt.Sprintf("%s-%s", strings.ReplaceAll(p.cfg.Identifier, ".", "-"), r.ID)
```

Comment:
This change has nothing to do with the topic of this PR; it is just there to make the linter happy.